vpr/src/route/connection_box_lookahead_map.cpp

```diff
@@ -497,27 +523,215 @@ static void run_dijkstra(int start_node_ind,
+                    if (!point.samples.empty()) {
+                        point.order = point.samples[0];
+                        sample_points.push_back(point);
+                    }
+                }
+            }
+        }
+    }
+
+    // sort by VPR coordinate
+    std::sort(sample_points.begin(), sample_points.end(),
+              [](const SamplePoint& a, const SamplePoint& b) {
+                  return a.order < b.order;
+              });
+
+    /* free previous delay map and allocate new one */
+    auto& device_ctx = g_vpr_ctx.device();
+    cost_map_.set_counts(segment_inf.size(),
+                         device_ctx.connection_boxes.num_connection_box_types());
+
+    VTR_ASSERT(REPRESENTATIVE_ENTRY_METHOD == SMALLEST);
+    RoutingCosts all_costs;
+
+    /* run Dijkstra's algorithm for each segment type & channel type combination */
+#if defined(VPR_USE_TBB)
+    tbb::mutex all_costs_mutex;
+    tbb::parallel_for_each(sample_points, [&](const SamplePoint& point) {
+#else
+    for (const auto& point : sample_points) {
+#endif
+        float max_cost = 0.f;
+        RoutingCosts costs;
+        SimpleCache<CompressedRoutingCostKey, float, DIJKSTRA_CACHE_SIZE> cache;
+        for (auto node_ind : point.samples) {
+            max_cost = std::max(max_cost, run_dijkstra(node_ind, &costs, &cache, COST_LIMIT));
+        }
+
+#if defined(VPR_USE_TBB)
+        all_costs_mutex.lock();
+#endif
+
+        VTR_LOG("Expanded sample point (%d, %d) %e miss %g\n",
+                point.location.x(), point.location.y(), max_cost, cache.miss_ratio());
+
+        // combine the cost map from this run with the final cost maps for each segment
+        for (const auto& cost : costs) {
+            const auto& key = cost.first;
+            const auto& val = cost.second;
+            const auto& x = all_costs.find(key);
+
+            // implements REPRESENTATIVE_ENTRY_METHOD == SMALLEST
+            if (x == all_costs.end() || x->second.cost_entry.delay > val.cost_entry.delay) {
+                all_costs[key] = val;
+            }
+        }
+
+#if defined(VPR_USE_TBB)
+        all_costs_mutex.unlock();
+    });
+#else
+    }
+#endif
+
+    VTR_LOG("Combining results\n");
+    /* boil down the cost list in routing_cost_map at each coordinate to a
+     * representative cost entry and store it in the lookahead cost map */
+    cost_map_.set_cost_map(all_costs);
+
+// diagnostics
+#if 0
+    for (auto &cost : all_costs) {
+        const auto& key = cost.first;
+        const auto& val = cost.second;
+        VTR_LOG("%d -> %d (%d, %d): %g, %g\n",
+                val.from_node, val.to_node,
+                key.delta.x(), key.delta.y(),
+                val.cost_entry.delay, val.cost_entry.congestion);
+    }
+#endif
+
+#if 1
+    for (int iseg = 0; iseg < (ssize_t)num_segments; iseg++) {
+        VTR_LOG("cost map for %s(%d)\n",
+                segment_inf[iseg].name.c_str(), iseg);
+        cost_map_.print(iseg);
+    }
+#endif
+
+#if 0
```

Comment:
For `if 0/1` it would be good to have a comment on why you might want to uncomment this section.  Maybe giving each block a descriptive `ifdef`?

Reply:
Done.